### ISSUE TITLE
torch.compile scatter_(reduce=) returns incorrect values on grouped_mm offs, MoE accumulate, and Llama 4 / Qwen3 router maps

### ISSUE THREAD

**Issue by KanakaPathivada** [torch-spyre/torch-spyre#4407]
### What happened?

`torch.compile` of `dst.scatter_(dim, index, src, reduce=...)` returns a tensor whose values disagree with CPU (`assert_close`, atol=rtol=0.01).

The kernel **runs** (no Inductor crash). `dynamic=False` and `dynamic=True` match each other and both disagree with CPU.

This is **`aten::scatter.reduce_out`** (legacy `reduce="add"` / `reduce="multiply"` with a tensor src). Models do not call this API; repro uses the **same dest ranks** as Gemma grouped_mm `scatter_add_` and Llama 4 / Qwen3 `torch.scatter` (`torch.manual_seed(0)`). Dest for multiply starts at ones so the product is not stuck at zero.

| pattern | reduce | dest | mismatch |
|---|---|---|---|
| grouped_mm offs | `multiply` fp16 | `[128]`, T=192 | **94 / 128 (73.4%)** — abs **inf** |
| grouped_mm offs | `add` int64 | `[128]`, T=192 | **95 / 128 (74.2%)** — abs **6** |
| MoE accumulate | `multiply` fp16 | `(8, 32, 64)` | **12288 / 16384 (75.0%)** — abs **inf** |
| MoE accumulate | `add` int64 | `(8, 32, 64)` | **12288 / 16384 (75.0%)** — abs **179** |
| Llama 4 ranks dim=1 K=1 | `multiply` fp16 | `(192, 128)` | **191 / 24576 (0.8%)** — abs **27.3** |
| Qwen3 ranks dim=1 K=8 | `multiply` fp16 | `(192, 128)` | **1536 / 24576 (6.2%)** — abs **0.168** |

Unique dest `(4, 64)` multiply is also wrong (**127 / 256 (49.6%)**). Not collision-only. Not fp16 rounding.

```
Tensor-likes are not close!

Mismatched elements: 95 / 128 (74.2%)
Greatest absolute difference: 6 at index (0,) (up to 0.01 allowed)
```

### What did you expect to happen?

Compiled Spyre should match CPU (`assert_close`, atol=rtol=0.01).

```python
import torch

torch.manual_seed(0)
device = torch.device("spyre:0")

G, T = 128, 192
dst = torch.ones(G, dtype=torch.int64)
idx = torch.randint(0, G, (T,), dtype=torch.int64)
src = torch.ones(T, dtype=torch.int64)

def scatter_add_red(d, i, s):
    out = d.clone()
    out.scatter_(0, i, s, reduce="add")
    return out

ref = scatter_add_red(dst, idx, src)
got = torch.compile(scatter_add_red, dynamic=False)(
    dst.to(device), idx.to(device), src.to(device)
).cpu()
torch.testing.assert_close(got, ref, atol=0.01, rtol=0.01)
```

### How can we reproduce it?

**Reproducer script:** [scatter_reductions_mismatch.py](https://github.com/user-attachments/files/32006376/scatter_reductions_mismatch.py)

```python
import torch

torch.manual_seed(0)
device = torch.device("spyre:0")

G, T = 128, 192
dst = torch.ones(G, dtype=torch.int64)
idx = torch.randint(0, G, (T,), dtype=torch.int64)
src = torch.ones(T, dtype=torch.int64)

def scatter_add_red(d, i, s):
    out = d.clone()
    out.scatter_(0, i, s, reduce="add")
    return out

ref = scatter_add_red(dst, idx, src)
got = torch.compile(scatter_add_red, dynamic=False)(
    dst.to(device), idx.to(device), src.to(device)
).cpu()
torch.testing.assert_close(got, ref, atol=0.01, rtol=0.01)
```

Same script: fp16 `reduce="multiply"` on those ranks, MoE dest `(8, 32, 64)`, Llama 4 `(192, 128)` K=1 dim=1, Qwen3 K=8.

### Any environmental details we need to know?

| | |
|-|-|
| Python | 3.12 |
| PyTorch | 2.13.0 |
| torch-spyre | commit - https://github.com/torch-spyre/torch-spyre/commit/8ef93123d6917a7b951768980b946064694d77d8 |
| Mode | `torch.compile(dynamic=False)` and `dynamic=True` |

### Relevant log output

```Shell
python scatter_reductions_mismatch.py
cpu = expected. compiled uses assert_close (atol=rtol=0.01).
aten::scatter.reduce_out compile mismatch. model ranks first.

--- 1. grouped_mm offs (G=128, T=192, 1-D) ---
[grouped_mm multiply] compile dest (128,) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  94 / 128 (73.4%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 94 / 128 (73.4%)
Greatest absolute difference: inf at index (0,) (up to 0.01 allowed)
Greatest relative difference: nan at index (0,) (up to 0.01 allowed)


[grouped_mm multiply] compile_dyn dest (128,) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  94 / 128 (73.4%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 94 / 128 (73.4%)
Greatest absolute difference: inf at index (0,) (up to 0.01 allowed)
Greatest relative difference: nan at index (0,) (up to 0.01 allowed)


[grouped_mm int64 add] compile dest (128,) torch.int64
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  95 / 128 (74.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 95 / 128 (74.2%)
Greatest absolute difference: 6 at index (0,) (up to 0.01 allowed)
Greatest relative difference: 0.8571428656578064 at index (0,) (up to 0.01 allowed)


[grouped_mm int64 add] compile_dyn dest (128,) torch.int64
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  95 / 128 (74.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 95 / 128 (74.2%)
Greatest absolute difference: 6 at index (0,) (up to 0.01 allowed)
Greatest relative difference: 0.8571428656578064 at index (0,) (up to 0.01 allowed)


--- 2. MoE accumulate (E=8, T=16, 32×64, 1-D ids expanded) ---
[MoE multiply expanded expert ids] compile dest (8, 32, 64) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  12288 / 16384 (75.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 12288 / 16384 (75.0%)
Greatest absolute difference: inf at index (0, 0, 0) (up to 0.01 allowed)
Greatest relative difference: nan at index (0, 0, 0) (up to 0.01 allowed)


[MoE multiply expanded expert ids] compile_dyn dest (8, 32, 64) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  12288 / 16384 (75.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 12288 / 16384 (75.0%)
Greatest absolute difference: inf at index (0, 0, 0) (up to 0.01 allowed)
Greatest relative difference: nan at index (0, 0, 0) (up to 0.01 allowed)


[MoE int64 add expanded expert ids] compile dest (8, 32, 64) torch.int64
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  12288 / 16384 (75.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 12288 / 16384 (75.0%)
Greatest absolute difference: 179 at index (0, 10, 45) (up to 0.01 allowed)
Greatest relative difference: 0.9944444298744202 at index (0, 10, 45) (up to 0.01 allowed)


[MoE int64 add expanded expert ids] compile_dyn dest (8, 32, 64) torch.int64
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  12288 / 16384 (75.0%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 12288 / 16384 (75.0%)
Greatest absolute difference: 179 at index (0, 10, 45) (up to 0.01 allowed)
Greatest relative difference: 0.9944444298744202 at index (0, 10, 45) (up to 0.01 allowed)


--- 3. Llama 4 ranks dim=1 dest (192, 128) K=1 multiply ---
[Llama4 multiply] compile dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  191 / 24576 (0.8%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 191 / 24576 (0.8%)
Greatest absolute difference: 27.28125 at index (191, 118) (up to 0.01 allowed)
Greatest relative difference: 0.96484375 at index (191, 118) (up to 0.01 allowed)


[Llama4 multiply] compile_dyn dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  191 / 24576 (0.8%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 191 / 24576 (0.8%)
Greatest absolute difference: 27.28125 at index (191, 118) (up to 0.01 allowed)
Greatest relative difference: 0.96484375 at index (191, 118) (up to 0.01 allowed)


--- 4. Qwen3 ranks dim=1 dest (192, 128) K=8 multiply ---
[Qwen3 multiply] compile dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile  MISMATCH  1536 / 24576 (6.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 1536 / 24576 (6.2%)
Greatest absolute difference: 0.16796875 at index (0, 49) (up to 0.01 allowed)
Greatest relative difference: 0.143798828125 at index (0, 49) (up to 0.01 allowed)


[Qwen3 multiply] compile_dyn dest (192, 128) torch.float16
[WARNING] [spyre.inductor.scratchpad.allocator] LX relayout is not supported by _make_cpsat_solver; continuing without relayout
  compile_dyn  MISMATCH  1536 / 24576 (6.2%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 1536 / 24576 (6.2%)
Greatest absolute difference: 0.16796875 at index (0, 49) (up to 0.01 allowed)
Greatest relative difference: 0.143798828125 at index (0, 49) (up to 0.01 allowed)


--- 5. unique dest control (4, 64) multiply ---
[fp16 multiply dim=0 unique] compile dest (4, 64) torch.float16
  compile  MISMATCH  127 / 256 (49.6%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 127 / 256 (49.6%)
Greatest absolute difference: 1198.0 at index (2, 63) (up to 0.01 allowed)
Greatest relative difference: 0.97705078125 at index (2, 62) (up to 0.01 allowed)


[fp16 multiply dim=0 unique] compile_dyn dest (4, 64) torch.float16
  compile_dyn  MISMATCH  127 / 256 (49.6%)
compiled spyre <-> cpu mismatch

Tensor-likes are not close!

Mismatched elements: 127 / 256 (49.6%)
Greatest absolute difference: 1198.0 at index (2, 63) (up to 0.01 allowed)
Greatest relative difference: 0.97705078125 at index (2, 62) (up to 0.01 allowed)
```
